### PR TITLE
rgw/cloud: Add custom headers for objects transitioned to cloud

### PR DESCRIPTION
**Pull request by soumyakoduri**
Some of the cloud services, (like MCG Noobaa/Azure Namespace store), may not be able to map and store the objects which contain s3 style metadata keys.

To help such services determine if the objects being transitioned are from RGW and whether or not ignore such attrs, added below two headers for the objects being copied/transitioned from RGW -

1) x-rgw-cloud : true/false
(set to "true" if the object is being transitioned/synced from RGW)

2) x-rgw-cloud-keep-attrs : true/false
- if set to default "true" , the cloud service should store all the x-amz-meta-rgwx-* attrs. If cannot be mapped/stored, the operation should fail
- if set to "false", the destination cloud can ignore such attrs and just store the object data being sent.

Also fixed a bug in the cloudtier module wherein the user-defined attrs were not being copied to the cloud endpoint as part of transition

Signed-off-by: Soumya Koduri <skoduri@redhat.com>
Fixes: https://tracker.ceph.com/issues/57980
Fixes:https://tracker.ceph.com/issues/58796

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by soumyakoduri** [comment]
@cbodley @mattbenjamin @yehudasa please let me know if these new headers added seem ok. If yes will add update the document https://docs.ceph.com/en/latest/radosgw/cloud-transition/#object-modification-limitations as well.

Also shall I add these headers for cloud synced objects as well?

**Comment by cbodley** [comment]
> 2) x-rgw-cloud-keep-rgwx-attrs : true/false
> - if set to default "true" , the cloud service should store all the x-amz-meta-rgwx-* attrs. If cannot be mapped/stored, the operation should fail

is there a reason to distinguish between `x-amz-meta-rgwx-` and any other `x-amz-meta-` attributes? doesn't cloud tiering expect all of that [user-defined object metadata](https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html#UserMetadata) to be mapped/preserved?

**Comment by yehudasa** [comment]
@cbodley are these even user visible attributes, or ones that we keep internally?

**Comment by cbodley** [comment]
> @cbodley are these even user visible attributes, or ones that we keep internally?

these rgwx attrs aren't stored in rgw's objects, but cloud transition is adding them to the objects we write to remote s3:
```
x-amz-meta-rgwx-versioned-epoch
x-amz-meta-rgwx-source-mtime
x-amz-meta-rgwx-source-etag
x-amz-meta-rgwx-source-key
x-amz-meta-rgwx-source-version-id
```
it doesn't look like `RGWLCCloudStreamPut::init_send_attrs()` is forwarding any x-amz-meta-* metadata stored in the rgw object, but i think it should. and i would expect noobaa not to mangle those metadata attributes either

the idea is that we should be able to someday transition these objects back to rgw in their original form

**Comment by soumyakoduri** [comment]
> > @cbodley are these even user visible attributes, or ones that we keep internally?
> 
> these rgwx attrs aren't stored in rgw's objects, but cloud transition is adding them to the objects we write to remote s3:
> 
> ```
> x-amz-meta-rgwx-versioned-epoch
> x-amz-meta-rgwx-source-mtime
> x-amz-meta-rgwx-source-etag
> x-amz-meta-rgwx-source-key
> x-amz-meta-rgwx-source-version-id
> ```
> 
> it doesn't look like `RGWLCCloudStreamPut::init_send_attrs()` is forwarding any x-amz-meta-* metadata stored in the rgw object, but i think it should. and i would expect noobaa not to mangle those metadata attributes either
> 
> the idea is that we should be able to someday transition these objects back to rgw in their original form


Right! Cloud tier should not be skipping any original object attrs. I will test that case and update. Thanks!

**Comment by cbodley** [comment]
> Right! Cloud tier should not be skipping any original object attrs. I will test that case and update. Thanks!

@soumyakoduri i don't think that's a blocker for progress here; maybe just drop the `rgwx-` part of `x-rgw-cloud-keep-rgwx-attrs` header and document that it covers the mapping of all `x-amz-meta-` attrs?

**Comment by soumyakoduri** [comment]
> > Right! Cloud tier should not be skipping any original object attrs. I will test that case and update. Thanks!
> 
> @soumyakoduri i don't think that's a blocker for progress here; maybe just drop the `rgwx-` part of `x-rgw-cloud-keep-rgwx-attrs` header and document that it covers the mapping of all `x-amz-meta-` attrs?

Done. I even addressed this bug of not copying "x-amz-" attrs in this PR. Please review.

File: doc/radosgw/cloud-transition.rst [hunk 1]
```diff
@@ -302,6 +302,18 @@ Due to API limitations there is no way to preserve original object modification
    x-amz-meta-rgwx-source-mtime: 1608546349.757100363
    x-amz-meta-rgwx-versioned-epoch: 0
```
Comment (anthonyeleven) [inline doc/radosgw/cloud-transition.rst:304]: Maybe expand attrs to attributes?
And s/all the//

**Review by anthonyeleven — COMMENTED** [review]
Docs suggestions

File: doc/radosgw/cloud-transition.rst (hunk 1 above)
Comment (soumyakoduri) [reply doc/radosgw/cloud-transition.rst:304]: Done. Thanks! Please review.

File: src/rgw/driver/rados/rgw_lc_tier.cc [hunk 2]
```diff
@@ -531,6 +530,12 @@ void RGWLCCloudStreamPut::init_send_attrs(const DoutPrefixProvider *dpp,
   for (auto& hi : rest_obj.attrs) {
     if (keep_attr(hi.first)) {
       attrs.insert(hi);
+    } else {
+      std::string s1 = boost::algorithm::to_lower_copy(hi.first);
+      const char* k = std::strstr(s1.c_str(), "x-amz");
```
Comment (cbodley) [inline src/rgw/driver/rados/rgw_lc_tier.cc:535]: there's a [boost::algorithm::istarts_with()](https://www.boost.org/doc/libs/1_79_0/doc/html/boost/algorithm/istarts_with.html) that looks perfect for this

File: doc/radosgw/cloud-transition.rst [hunk 3]
```diff
@@ -302,6 +302,18 @@ Due to API limitations there is no way to preserve original object modification
...
 
+In order to allow some cloud services detect the source and map the user-defined 'x-amz-meta-' attributes, below two additional new attributes are added to the objects being transitioned 
+
+::
+    
+   x-rgw-cloud : true/false
+   (set to "true", by default, if the object is being transitioned from RGW)
+
+   x-rgw-cloud-keep-rgwx-attrs : true/false
```
Comment (cbodley) [inline doc/radosgw/cloud-transition.rst:312]: x-rgw-cloud-keep-attrs

File: src/rgw/driver/rados/rgw_lc_tier.cc (hunk 2 above)
Comment (soumyakoduri) [reply src/rgw/driver/rados/rgw_lc_tier.cc:535]: I observed that user defined attrs are stored with extra 'user-' prefix in RGW (for eg., user-x-amz-meta-key1). So I  used std::strstr here to split and fetch only the original key (x-amz-meta-key1) to be added as header.

File: doc/radosgw/cloud-transition.rst (hunk 3 above)
Comment (soumyakoduri) [reply doc/radosgw/cloud-transition.rst:312]: Done.

**Review by cbodley — APPROVED** [review]

**Comment by soumyakoduri** [comment]
jenkins render docs

**Comment by soumyakoduri** [comment]
http://pulpito.front.sepia.ceph.com/soumyakoduri-2023-02-22_06:56:10-rgw:cloud-transition-wip-skoduri-cloud-trans-azure-distro-default-smithi

http://pulpito.front.sepia.ceph.com/soumyakoduri-2023-02-22_06:56:26-rgw-wip-skoduri-cloud-trans-azure-distro-default-smithi

**Comment by cbodley** [comment]
> http://pulpito.front.sepia.ceph.com/soumyakoduri-2023-02-22_06:56:10-rgw:cloud-transition-wip-skoduri-cloud-trans-azure-distro-default-smithi

👍 

> 
> http://pulpito.front.sepia.ceph.com/soumyakoduri-2023-02-22_06:56:26-rgw-wip-skoduri-cloud-trans-azure-distro-default-smithi

the full suite might take a while. is there any interesting test coverage we're missing from the rgw/cloud-transition suite above? if not, i'm fine with merging based on that

**Comment by soumyakoduri** [comment]
> p.s. please use the public "pulpito.ceph.com" links instead of "pulpito.front.sepia.ceph.com" so people without vpn access can view the results!

Thanks Casey! Updated the links.

**Pull request merged**